db/sync.go

```diff
@@ -52,10 +52,13 @@ func (db *DB) BeginSync(batchPushURL string, diffServerURL string, dataLayerAuth
 	}
 	newSnapshot, clientViewInfo, err := db.puller.Pull(db.noms, headSnapshot, diffServerURL, dataLayerAuth, db.clientID)
 	if err != nil {
-		return hash.Hash{}, syncInfo, fmt.Errorf("sync failed: %w", err)
+		return hash.Hash{}, syncInfo, fmt.Errorf("sync failed: pull from %s failed: %w", diffServerURL, err)
 	} else {
 		syncInfo.ClientViewInfo = clientViewInfo
 	}
+	if newSnapshot.Meta.Snapshot.ServerStateID == headSnapshot.Meta.Snapshot.ServerStateID {
+		return hash.Hash{}, syncInfo, fmt.Errorf("sync failed: no new data (client and server both have %s)", headSnapshot.Meta.Snapshot.ServerStateID)
```

Comment:
Why is this an error? It's going to be very common for it to happen, right? Can it just be a no-op, but not an error?

Reply:
was trying to make it simple to understand the signature: no error -> continue sync protocol, error -> do not continue. the obvious alternative is to return on error only on something truly unexpected and have the presence or absence of synchead indicate whether to continue the protocol. will leave this for my list.

btw we should continue to keep the option in the back of our minds to model the repm api  more like an rpc where you get a fooresult back that has everything you need eg whether it was a success, error message, and any results (and not like a golang foo, error signature where they are separate). it's  easy to send that kind of thing across any kind of boundary and you dont have to worry about these kinds of questions too much. and you dont have to worry eg how to map it to http, or asynchronous messages, or whatever should you need to, you just send the request and result. bindings look more similar to each other which i guess has pros and cons.